firedantic/models.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABC
+from typing import List, Optional, Type, TypeVar
+
+import pydantic
+from firedantic.configurations import CONFIGURATIONS
+from firedantic.exceptions import ModelNotFoundError
+from google.cloud.firestore_v1 import CollectionReference, DocumentReference
+
+TModel = TypeVar("TModel", bound="Model")
+
+
+class Model(pydantic.BaseModel, ABC):
+    """Base model class.
+
+    Implements basic functionality for Pydantic models, such as save, delete, find etc.
+    """
+
+    __collection__: Optional[str] = None
+
+    id: Optional[str] = None
+
+    def save(self) -> None:
+        """Saves this model in the database."""
+        data = self.dict()
+        if "id" in data:
+            del data["id"]
+
+        doc_ref = self._get_doc_ref()
+        doc_ref.set(data)
+        self.id = doc_ref.id
+
+    def delete(self) -> None:
+        """Deletes this model from the database."""
+        self._get_doc_ref().delete()
+
+    @classmethod
+    def find(cls: Type[TModel], filter_: dict) -> List[TModel]:
+        """Returns a list of models from the database based on a filter.
+
+        :param filter_: The filter criteria.
+        :return: List of found models.
+        """
+        coll = cls._get_col_ref()
+
+        query = coll
+
+        for key, value in filter_.items():
+            query = query.where(key, "==", value)
+
+        return [cls(id=doc.id, **doc.to_dict()) for doc in query.stream()]
+
+    @classmethod
+    def find_one(cls: Type[TModel], filter_: dict) -> TModel:
+        """Returns one model from the DB based on a filter.
+
+        :param filter_: The filter criteria.
+        :return: The model instance.
+        :raise ModelNotFoundError: If the entry is not found.
+        """
+        try:
+            return cls.find(filter_)[0]
+        except IndexError:
+            raise ModelNotFoundError("Model not found")
```

Comment:
Should this error text be aligned a bit more with the one for `get_by_id`; something like `No '{cls.__name__}' found`?

Reply:
Sure, sounds good.